# Added examples for GraphScene to docs (#444)
ManimCommunity/manim: docs/source/examples/plots_examples.rst

Plotting with manim
=================================

Examples to illustrate the use of GrapheScenes in Manim

.. manim:: Plot1
    :quality: medium
    :save_last_frame:

    class Plot1(GraphScene):
        def construct(self):
            self.setup_axes()

>             my_func = lambda x: np.sin(x)
>             func_graph=self.get_graph(my_func)

Reviewer: ```suggestion
            func_graph=self.get_graph(lambda x: np.sin(x))
```

            self.add(func_graph)

.. manim:: Plot2yLabel
    :quality: medium
    :save_last_frame:

    class Plot2yLabel(GraphScene):
        CONFIG = {
            "y_min": 0,
            "y_max": 100,
            "y_tick_frequency": 10,
            "y_labeled_nums": np.arange(0, 100, 10)
        }

        def construct(self):
            self.setup_axes(animate=True)

>             dot = Dot()
>             dot.move_to(self.coords_to_point(PI / 2, 20))
>             my_func = lambda x: 20 * np.sin(x)
>             func_graph = self.get_graph(my_func)

Reviewer: ```suggestion
            dot = Dot().move_to(self.coords_to_point(PI / 2, 20))
            func_graph = self.get_graph(lambda x: 20 * np.sin(x))
```

            self.add(func_graph)
            self.add(dot)

.. manim:: Plot3DataPoints
    :quality: medium
    :save_last_frame:

    class Plot3DataPoints(GraphScene):
        CONFIG = {
            "y_axis_label": r"Concentration[$\%$]",
            "x_axis_label": "Time [s]",
            }

        def construct(self):
            data = [1, 2, 2, 4, 4, 1, 3]
            self.setup_axes(animate=True)
            for time, dat in enumerate(data):
                dot = Dot().move_to(self.coords_to_point(time, dat))
                self.add(dot)
                self.wait(1)

>                 print(time)

Reviewer: This won't make any difference at all in the shown example so I'd suggest we remove it.

Author: done

            self.wait()

.. manim:: Plot3bGaussian
    :quality: medium
    :save_last_frame:

    class Plot3bGaussian(GraphScene):
        def construct(self):
            def gaussian(x):
                mu=3; sig= 1; amp=5

>                 return amp*np.exp( ( -1/2 * ( (x-mu)/sig)**2 ) )

Reviewer: This looks weird to me, and it's usually not recommended to use the semi-colon in python if it can be avoided.

Why not define `gaussian` outside of the class? And mu, sig, amp could be global variables as well. Right now, it kinda gives the (wrong) impression that `gaussian` needs to be defined inside `construct`.

Author: Ok, I will define it outside.
Sometimes I prefer it when I don't have to carry this self. in the code.
When everything is in "construct", it feels a little tighter to me.
But I can add another example, one where a function is defined in construct, and another one where the function is defined outside of the construct. 
In my computer class, I learned that we should best never use any global variables whenever they are not necessary.

Reviewer: Woops sorry I missed this comment :smile: 

I agree that global constants are to be avoided (especially perhaps in python), but this seems like the right place for them. They are literally both *global* and *constant* (at least for the purposes of this example).

            self.setup_axes()

>             graph = self.get_graph(gaussian, x_min=-1, x_max=10, ).set_stroke(width=5)

Reviewer: ```suggestion
            graph = self.get_graph(gaussian, x_min=-1, x_max=10).set_stroke(width=5)
```

            func_graph2 = self.get_graph(func_sin)
            vert_line = self.get_vertical_line_to_graph(TAU, func_graph, color=YELLOW)
            graph_lab = self.get_graph_label(func_graph, label="\\cos(x)")
            graph_lab2 = self.get_graph_label(func_graph2, label="\\sin(x)", x_val=-10, direction=UP / 2)
            two_pi = TexMobject("x = 2 \\pi")
            label_coord = self.input_to_graph_point(TAU, func_graph)
            two_pi.next_to(label_coord, RIGHT + UP)
            self.add(func_graph, func_graph2, vert_line, graph_lab, graph_lab2, two_pi)

.. manim:: Plot5Area
    :quality: medium
    :save_last_frame:

    class Plot5Area(GraphScene):
        CONFIG = {
            "x_min" : 0,
            "x_max" : 5,
            "y_min" : 0,
            "y_max" : 6,
            "y_tick_frequency" : 1,
            "x_tick_frequency" : 1,
            "x_labeled_nums" : [0,2,3]
        }
        def construct(self):
            self.setup_axes(animate=False)
            curve1 = self.get_graph(lambda x : 4*x-x**2, x_min=0,x_max=4)
            curve2 = self.get_graph(lambda x : 0.8*x**2-3*x+4, x_min=0,x_max=4)

>             line1 = self.get_vertical_line_to_graph(2,curve1,DashedLine,color=YELLOW)
>             line2 = self.get_vertical_line_to_graph(3,curve1,DashedLine,color=YELLOW)
>             area1 = self.get_area(curve1,0.3,0.6, dx_scaling=10, area_color=BLUE)
>             area2 = self.get_area(curve2,2,3,bounded=curve1)
>             self.add(curve1,curve2,line1,line2,area1,area2)

Reviewer: ```suggestion
            line1 = self.get_vertical_line_to_graph(2, curve1, DashedLine, color=YELLOW)
            line2 = self.get_vertical_line_to_graph(3, curve1, DashedLine, color=YELLOW)
            area1 = self.get_area(curve1, 0.3, 0.6, dx_scaling=10, area_color=BLUE)
            area2 = self.get_area(curve2, 2, 3, bounded=curve1)
            self.add(curve1, curve2, line1, line2, area1, area2)
```

.. manim:: Plot6HeatDiagram
    :quality: medium
    :save_last_frame:

    class Plot6HeatDiagram(GraphScene):
        CONFIG = {
            "y_axis_label": r"T[$^\circ C$]",
            "x_axis_label": r"$\Delta Q$",
            "y_min": -8,
            "y_max": 30,
            "x_min": 0,
            "x_max": 40,
            "y_labeled_nums": np.arange(-5, 34, 5),
            "x_labeled_nums": np.arange(0, 40, 5),

        }

        def construct(self):
            data = [20, 0, 0, -5]
            x = [0, 8, 38, 39]

>             print(x)

Reviewer: I think we should remove any `print` from the examples.

Author: yes, you are totally right.

            self.setup_axes(animate=True)
            dot_collection = VGroup()
            for time, val in enumerate(data):
                dot = Dot().move_to(self.coords_to_point(x[time], val))
                self.add(dot)
                dot_collection.add(dot)
            l1 = Line(dot_collection[0].get_center(), dot_collection[1].get_center())
            l2 = Line(dot_collection[1].get_center(), dot_collection[2].get_center())
            l3 = Line(dot_collection[2].get_center(), dot_collection[3].get_center())
            self.add(l1, l2, l3)

> This is an parametric function

Reviewer: ```suggestion
This is a parametric function
```

Plotting with manim
=================================

> Examples to illustrate the use of GraphScenes in manim

Reviewer 2: ```suggestion
Examples to illustrate the use of :class:`.GraphScene` in manim.
```

.. manim:: Plot1
    :quality: medium
    :save_last_frame:

> 

Reviewer 3: **Nitpick :** Maybe add `import numpy as np` 

(here and in every example that uses numpy)

Reviewer 2: When doing `from manim import *`, numpy is also imported as `np`. At some point, we have decided not to show imports in the examples -- but we can certainly discuss this again.

Reviewer 3: Is numpy in __init__ ? 
But still, I think for good practice purpose we should add it.

Author: We can write at one point before the examples begin, that all examples are run with `from manim import *``and that this also includes `import numpy as np`

Reviewer: I kinda like examples that are fully copy&pastable. I think in the best case scenario, any user would be able to copy the example code and run it immediately in their system. So I'm in favor of explictly including the import statments.

Author: I prefer it to have the code as compact as possible so that you don't have to scroll extra long.
In this case, we only have 9 examples, that are with the import statement 18 extra lines of code.
But maybe we can discuss this in a separate issue.
Can we leave it for now without the import statement?
When it's consensus that they should be included, then I am ok with changing it in a future pr.

Reviewer: If the code is too long, then maybe each example should have its own page, like matplotlib does.

But yes let's leave that for another issue. I'm fine with merging this PR as is.

Author: nice!! 
See this question in of import statement in #454

Author: >  I'm fine with merging this PR as is.
@leotrs : could you then give it an approve? :)

    class Plot1(GraphScene):
        def construct(self):
            self.setup_axes()
            func_graph=self.get_graph(lambda x: np.sin(x))
            self.add(func_graph)

.. manim:: Plot2yLabel
    :quality: medium
    :save_last_frame:

    class Plot2yLabel(GraphScene):
        CONFIG = {
            "y_min": 0,
            "y_max": 100,

>             "y_tick_frequency": 10,

Reviewer 2: `y_tick_frequency` does not seem to exist, is it possible that this should be something like

```suggestion
            "y_axis_config": {"tick_frequency": 10},
```

?

            "y_labeled_nums": np.arange(0, 100, 10)
        }

        def construct(self):
            self.setup_axes(animate=True)
            dot = Dot().move_to(self.coords_to_point(PI / 2, 20))
            func_graph = self.get_graph(lambda x: 20 * np.sin(x))
            self.add(func_graph)
            self.add(dot)

.. manim:: Plot3DataPoints
    :quality: medium
    :save_last_frame:

    class Plot3DataPoints(GraphScene):
        CONFIG = {

>             "y_axis_label": r"Concentration[$\%$]",

Reviewer 2: Math mode should not be required to typeset %.

```suggestion
            "y_axis_label": r"Concentration[\%]",
```

            "x_axis_label": "Time [s]",
            }

        def construct(self):
            data = [1, 2, 2, 4, 4, 1, 3]
            self.setup_axes(animate=True)
            for time, dat in enumerate(data):
                dot = Dot().move_to(self.coords_to_point(time, dat))
                self.add(dot)

.. manim:: Plot3bGaussian
    :quality: medium
    :save_last_frame:

    amp = 5
    mu = 3
    sig = 1

    def gaussian(x):
        return amp * np.exp((-1 / 2 * ((x - mu) / sig) ** 2))

    class Plot3bGaussian(GraphScene):
        def construct(self):
            self.setup_axes()
            graph = self.get_graph(gaussian, x_min=-1, x_max=10).set_stroke(width=5)
            self.add(graph)

.. manim:: Plot3cGaussian
    :quality: medium
    :save_last_frame:

    class Plot3cGaussian(GraphScene):
        def construct(self):
            def gaussian(x):
                amp = 5
                mu = 3
                sig = 1

>                 return amp * np.exp((-1 / 2 * ((x - mu) / sig) ** 2))

Reviewer 2: `Plot3bGaussian` and `Plot3cGaussian` should use the same strategy w.r.t. `gaussian` and `amp, mu, sig`.

Author: In this case, b and c have the purpose of showing that is possible on the one hand to define math functions outside, and on the other hand inside the construct class.
One is the more pythonic way, the other way is the tighter and more convenient way in my opinion.
Here, it just illustrates that both is possible.
In a future PR, I can write some things to the purpose of these examples.

Reviewer: I prefer to have `gaussian` outside `construct` (and outside `Scene`, really) bc it seems more pythonic. But I agree it should be illustrated that both are possible. Not sure if two videos/images are necessary for this tho, maybe just include one and mention that the alternative is possible?

Author: Most times as a reader, I would prefer two explicit examples over one example and the sentence "and you can also do this"

Reviewer: I think it's fine to do that, especially if we're trying to encourage one and not the other. If one is more pythonic, I think we should encourage that one.

"There should be one-- and preferably only one --obvious way to do it."

Reviewer: > Even if the code looks not perfect, what matters is the resulting video and not that it is coded in the way that is to a most common convention.

I don't care for perfect code, I care for consistency. Whatever the consensus is, we should stick to it as much as possible.

Author: For me, manim has a very strong advantage that you can do things in lots of different ways, depending on your current situation.
With this example, I want to showcase that we have here two convenient ways of doing things.
I would be fine with adding one line like:
`Here is a script that shows how to define a function in a most pythonic way`
and to the other one:
`here is a script that gives you the opportunity to define your function close to where it's needed, which might be useful in the case of long scripts.`

Reviewer: I think part of my hesitation here is that these two examples are really illustrating a python feature, not a manim feature. But I'll stop now and let you and others decide :)

Reviewer 2: In principle, I agree with @leotrs: we should primarily showcase manim features over python features.

Anyhow, I think we can still merge this as-is. I am pretty sure that we will revisit these examples sooner or later to add some more explanations and stuff -- and then we might even have a better idea of what precisely should be illustrated with these examples.

Reviewer: Agreed, let's merge as is and continue iterating rapidly :)

            self.setup_axes()
            graph = self.get_graph(gaussian, x_min=-1, x_max=10).set_style(stroke_width=5, stroke_color=GREEN)
            self.add(graph)

.. manim:: Plot4SinCos
    :quality: medium
    :save_last_frame:

    class Plot4SinCos(GraphScene):
        CONFIG = {
            "x_min": -10,
            "x_max": 10.3,
            "num_graph_anchor_points": 100,
            "y_min": -1.5,
            "y_max": 1.5,
            "graph_origin": ORIGIN,
            "function_color": RED,
            "axes_color": GREEN,
            "x_labeled_nums": range(-10, 12, 2),
        }

        def construct(self):
            self.setup_axes(animate=False)

            def func_cos(x):
                return np.cos(x)

            def func_sin(x):
                return np.sin(x)

            func_graph = self.get_graph(func_cos, self.function_color)
            func_graph2 = self.get_graph(func_sin)
            vert_line = self.get_vertical_line_to_graph(TAU, func_graph, color=YELLOW)
            graph_lab = self.get_graph_label(func_graph, label="\\cos(x)")
            graph_lab2 = self.get_graph_label(func_graph2, label="\\sin(x)", x_val=-10, direction=UP / 2)

>             two_pi = TexMobject("x = 2 \\pi")

Reviewer 2: `TexMobject` is deprecated, this should be `MathTex`:
```suggestion
            two_pi = MathTex(r"x = 2 \pi")
```

            self.add(curve1, curve2, line1, line2, area1, area2)

.. manim:: Plot6HeatDiagram
    :quality: medium
    :save_last_frame:

    class Plot6HeatDiagram(GraphScene):
        CONFIG = {
            "y_axis_label": r"T[$^\circ C$]",
            "x_axis_label": r"$\Delta Q$",
            "y_min": -8,
            "y_max": 30,
            "x_min": 0,
            "x_max": 40,
            "y_labeled_nums": np.arange(-5, 34, 5),
            "x_labeled_nums": np.arange(0, 40, 5),

        }

        def construct(self):
            data = [20, 0, 0, -5]
            x = [0, 8, 38, 39]
            self.setup_axes(animate=True)
            dot_collection = VGroup()
            for time, val in enumerate(data):
                dot = Dot().move_to(self.coords_to_point(x[time], val))
                self.add(dot)
                dot_collection.add(dot)
            l1 = Line(dot_collection[0].get_center(), dot_collection[1].get_center())
            l2 = Line(dot_collection[1].get_center(), dot_collection[2].get_center())
            l3 = Line(dot_collection[2].get_center(), dot_collection[3].get_center())
            self.add(l1, l2, l3)

> This is a parametric function

Reviewer 2: Just a suggestion for a small improvement to this (temporary) introduction to parametric plots.

```suggestion
The following example illustrates how to draw parametric function plots.
```
